x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java

```diff
@@ -794,316 +768,298 @@ private void findTokenFromRefreshToken(String refreshToken, ActionListener<Searc
-                executeAsyncWithOrigin(client.threadPool().getThreadContext(), SECURITY_ORIGIN, updateRequest.request(),
-                    ActionListener.<UpdateResponse>wrap(
-                        updateResponse -> {
-                            if (updateResponse.getResult() == DocWriteResponse.Result.UPDATED) {
-                                logger.debug("updated the original token document to {}", updateResponse.getGetResult().sourceAsMap());
-                                createUserToken(newUserTokenId, authentication, clientAuth, listener, metadata, true);
-                            } else if (backoff.hasNext()) {
-                                logger.info("failed to update the original token document [{}], the update result was [{}]. Retrying",
-                                    tokenDocId, updateResponse.getResult());
-                                final Runnable retryWithContextRunnable = client.threadPool().getThreadContext()
-                                        .preserveContext(() -> innerRefresh(tokenDocId, source, seqNo, primaryTerm, clientAuth, listener,
-                                                backoff, refreshRequested));
-                                client.threadPool().schedule(retryWithContextRunnable, backoff.next(), GENERIC);
-                            } else {
-                                logger.info("failed to update the original token document [{}] after all retries, " +
-                                    "the update result was [{}]. ", tokenDocId, updateResponse.getResult());
+            });
+        } else {
+            final String newUserTokenId = UUIDs.randomBase64UUID();
+            final Map<String, Object> updateMap = new HashMap<>();
+            updateMap.put("refreshed", true);
+            updateMap.put("refresh_time", clock.instant().toEpochMilli());
+            updateMap.put("superseded_by", getTokenDocumentId(newUserTokenId));
+            assert seqNo != SequenceNumbers.UNASSIGNED_SEQ_NO : "expected an assigned sequence number";
+            assert primaryTerm != SequenceNumbers.UNASSIGNED_PRIMARY_TERM : "expected an assigned primary term";
+            final UpdateRequestBuilder updateRequest = client.prepareUpdate(SECURITY_INDEX_NAME, SINGLE_MAPPING_NAME, tokenDocId)
+                    .setDoc("refresh_token", updateMap)
+                    .setFetchSource(true)
+                    .setRefreshPolicy(RefreshPolicy.IMMEDIATE)
+                    .setIfSeqNo(seqNo)
+                    .setIfPrimaryTerm(primaryTerm);
+            executeAsyncWithOrigin(client.threadPool().getThreadContext(), SECURITY_ORIGIN, updateRequest.request(),
+                    ActionListener.<UpdateResponse>wrap(updateResponse -> {
+                        if (updateResponse.getResult() == DocWriteResponse.Result.UPDATED) {
+                            logger.debug("updated the original token document to {}", updateResponse.getGetResult().sourceAsMap());
```

Comment:
the source as map is lazily computed so I think we should consider guarding this log statement with a `isDebugEnabled` check

Reply:
I have sprinkled `() -> new ParameterizedMessage(` in a couple of places, failure handlers that are not quite exceptional but retry-able.